### ISSUE TITLE
Improve JSON output

### ISSUE THREAD

**Issue by pmos69** [mavam/vast#98]
Is the json output standard?
Tested in http://jsonlint.com/# it gives errors, as well as trying to process with standard libraries.

The output is the form contains multiple records in the form: ```{ ...record1...}{...record2...}```, but AFAIK, the correct form for multiple records in a single json file whould be: ```[{...record1...},{...record2...}]``` (straight braces enclosing a comma delimited list of records).

There is also contents in the records that seems to deviate from the standard (not sure).
ex: I get the following in the middle of bro records:
```
 ...
},
    "type": "bro::conn = record {ts: time, ...
```
Is this correct?
(both the ```bro::conn = ...``` and the ```record``` construct that appears also in other places)

**Comment by mavam** [comment]
> Is the json output standard?
> Tested in http://jsonlint.com/# it gives errors, as well as trying to process with standard libraries.
> 
> The output is the form contains multiple records in the form: { ...record1...}{...record2...}, but AFAIK, the correct form for multiple records in a single json file whould be: [{...record1...},{...record2...}] (straight braces enclosing a comma delimited list of records).

You're hitting the pain point of the current setup: the lack of framing for event boundaries. The JSON sink displays events on standard output, one at a time, concatenated by newlines. I think this is what you're pointing out. Since it's a finite but potentially very long stream, I wasn't sure how to solve this scenario, so I defaulted to the Bro approach, which concatenates events with new lines:

    bro -r ~/code/vast/brocon/pcap/2009-M57-day11-18.trace -e 'redef LogAscii::use_json = T;'

produces 

```json
{"ts":1258568537.026452,"uid":"Ctfm7h38JyGhfYCY6a","id.orig_h":"192.168.1.102","id.orig_p":1649,"id.resp_h":"198.189.255.83","id.resp_p":80,"proto":"tcp","service":"http","duration":19.812889,"orig_bytes":4134,"resp_bytes":33021,"conn_state":"S1","missed_bytes":0,"history":"ShADad","orig_pkts":21,"orig_ip_bytes":4982,"resp_pkts":30,"resp_ip_bytes":34229,"tunnel_parents":[]}
{"ts":1258568324.7392,"uid":"CyXUnd2rSg487uWySl","id.orig_h":"192.168.1.105","id.orig_p":49518,"id.resp_h":"198.189.255.75","id.resp_p":80,"proto":"tcp","service":"http","duration":0.242349,"orig_bytes":1440,"resp_bytes":14579,"conn_state":"S1","missed_bytes":0,"history":"ShADad","orig_pkts":13,"orig_ip_bytes":1972,"resp_pkts":14,"resp_ip_bytes":15151,"tunnel_parents":[]}
{"ts":1258568420.922323,"uid":"CFuwxq3Yv9qfDaIef7","id.orig_h":"192.168.1.105","id.orig_p":49555,"id.resp_h":"198.105.194.142","id.resp_p":80,"proto":"tcp","service":"http","duration":0.162659,"orig_bytes":1292,"resp_bytes":134,"conn_state":"S1","missed_bytes":0,"history":"ShADad","orig_pkts":5,"orig_ip_bytes":1504,"resp_pkts":4,"resp_ip_bytes":298,"tunnel_parents":[]}
```

The linter will fail for this output for the same reason as VAST. Should we switch to a "half-open" array instead, as you hint? Then we'd have `[e0, e1, ...` and finish with the closing `]` as soon as the full result is there. The downside would be that the full result would have to be there for proper JSON. But right now it's not proper either.

>     "type": "bro::conn = record {ts: time, ...
> Is this correct?
> (both the bro::conn = ... and the record construct that appears also in other places

What you're seeing here are the type information. VAST prints events as a JSON object with meta data keys (`id` and `timestamp`) and the value, which itself is an object with keys `data` and `type`. To enable unambiguous data representation, each event contains the type information as ASCII string separately, which is the `bro::conn = record { ...` you're seeing.

**Comment by pmos69** [comment]
But instead of:
```
 ...
},
    "type": "bro::intel = record {ts: time, uid: ...
```

shouldn't it be something like:
```
 },
    "type": "bro::intel",
    "types": {ts: time, uid: ...}
```
That representation would be json compliant.

Also in other places. ex: ```id: record {orig_h: addr, orig_p: port,...}``` if represented as ```id: {orig_h: addr, orig_p: port,...}``` would be compliant, without information loss.

If the output was json compliant, a lot of libs and applications could ingest it.
If the current output is required, perhaps another, standard based, json sink could be added for compatibility reasons.
The output could also be flatter, without loss of information, and that could be good.

**Comment by pmos69** [comment]
I think you already have flattening functions in your code, so this might be more or less straightforward...

**Comment by mavam** [comment]
> shouldn't it be something like:
> 
>  },
>     "type": "bro::intel",
>     "types": {ts: time, uid: ...}

Yes, this would certainly be nicer. Right now the type information is a plain string, something that applications cannot readily work with and have to parse. Because I didn't have a concrete use case yet, I kept it as an opaque string. 

The proper approach would be to make the class `vast::type` convertible to JSON. Internally, VAST uses the *Convertible* concept for this, so the change would mean simply providing the overload `bool convert(type const& t, json& j)`. Then we can use `to_json(type const&)` to render the type in a more structured manner.

> Also in other places. ex: id: record {orig_h: addr, orig_p: port,...} if represented as id: {orig_h: addr, orig_p: port,...} would be compliant, without information loss.

One thing to note: the type string is *not* JSON, even though it looks very similar. It's just a plain string in VAST's own type language (derived from Bro) used in schemas. 

> If the current output is required, perhaps another, standard based, json sink could be added for compatibility reasons.

I find two JSON sinks confusing. Let's make the one we have standard-compliant.

> The output could also be flatter, without loss of information, and that could be good.

By this, do you mean that the output should not be in tree structure but rather in one line per event?

To summarize, the action items are:

1. Wrap events into an array as opposed to separating them by newlines
2. Write the type information out as JSON as opposed to a plain string.
3. Add an option so that users can flatten records.

Anything else?

**Comment by pmos69** [comment]
>>  Also in other places. ex: id: record {orig_h: addr, orig_p: port,...} if represented as id: {orig_h: addr, orig_p: port,...} would be compliant, without information loss.

>One thing to note: the type string is not JSON, even though it looks very similar. It's just a plain string in VAST's own type language (derived from Bro) used in schemas.

I see it now.
In ```"type": "bro::intel = record {ts: time, uid: string, id: record {orig_h: addr, orig_p: port, resp_h: addr, resp_p: port}, fuid: string, file_mime_type: string, file_desc: string, seen: record {indicator: string, indicator_type: string, where: string, node: string}, sources: set<string>}"``` the the key is ```type``` and the value is a string (```"bro::intel = record {ts: ... }"```).
Got me there :) , but pitty that info is not comverted to JSON as well.

>>If the current output is required, perhaps another, standard based, json sink could be added for compatibility reasons.

>I find two JSON sinks confusing. Let's make the one we have standard-compliant.

I agree.
>>The output could also be flatter, without loss of information, and that could be good.

>By this, do you mean that the output should not be in tree structure but rather in one line per event?

No, not that. JSON multi-line trees are no problem.
It's the multi level JSON structure itself that can complicate some json operations. You have a flatting function for records in the code. Saw it when browsing the tests in ```test/unit/tests/type.cc```.
Couldn't that be applied to the record prior to the JSON output? (as an option)

**Comment by mavam** [comment]
> Couldn't that be applied to the record prior to the JSON output? (as an option)

Ah, I see. That makes sense. Then instead of nested JSON objects representing nested records, we'd use a single JSON object with keys akin to Bro logs (`id.orig_h` for `orig_h` within the record named `id`). The flattening transformation is straight-forward, so I don't see anything that would prevent that.

Do you have a use case for that (i.e., an application that can only handle flat keys) or were you thinking this is just a nice-to-have feature?

**Comment by pmos69** [comment]
I opened the issue after trying to manipulate Vast JSON output with python JSON libraries, like the built in JSON library and the panda library (very useful for a lot of data-frame manipulation tasks and calculations).
Anyway, about the flat level JSON, it just makes it easier to deal with the data in some cases, especially when trying to deal with JSON records with different structures within the same dataset.
With the panda library, for example, I had some issues trying to manipulate deep nested fields. (I "massaged" some Vast JSON output to make it JSON compliant before handling it).

**Comment by mavam** [comment]
> Wrap events into an array as opposed to separating them by newlines

This is now implemented in `topic/json-tweaks`. Two more items remaining:

- [ ] Write the type information out as JSON as opposed to a plain string.
- [ ] Add an option so that users can flatten records.
